src/write.c

```diff
@@ -1325,6 +1326,48 @@ static avifResult avifImageCreateAllocate(avifImage ** sampleTransformedImage, c
     return avifImageAllocatePlanes(*sampleTransformedImage, planes);
 }
 
+static avifResult avifEncoderDecodeSatoBaseImage(const avifEncoder * encoder,
+                                                 const avifImage * original,
+                                                 uint32_t numBits,
+                                                 avifPlanesFlag planes,
+                                                 avifCodec ** codec,
```

Comment:
It is better to make `codec` a local variable in this function instead of an in/out parameter. This function needs to call `avifCodecDestroy(codec)` whether it succeeds or fails. This means we won't be able to use the `AVIF_CHECKERR` macro after `codec` has been created and before it is destroyed.

Reply:
That was the point: to be able to use `AVIF_CHECKRES()` within `avifEncoderDecodeSatoBaseImage()` and just having to handle `avifImageDestroy(decodedBaseImage)` and `avifCodecDestroy(codec)` in a single location, at its call site.

I added a comment to `avifEncoderDecodeSatoBaseImage()`.

Reply:
In C code this kind of resource cleanup is best handled by `goto` statements. (This is an acceptable use of `goto`.)

I understand the debugger break in the `AVIF_CHECKRES()` macro is helpful, but trying to use `AVIF_CHECKRES()` in this function leaks the implementation detail to the caller. This is also why `(*codec)->diag` cannot point to a local variable, and we had to remove `const` from the `avifEncoder * encoder` parameters of three functions in the call chain.

Reply:
I just realized that for better error reporting we should set `(*codec)->diag` to `&encoder->diag`, so we have to remove `const` from the `avifEncoder * encoder` parameters of those three functions in any case.